keps/sig-api-machinery/0014-garbage-collection.md

```diff
@@ -0,0 +1,555 @@
+state, the TTL will be reset. The resource will only be deleted when it is a
+garbage collection candidate and its TTL has expired (zero TTL expires
+immediately). Garbage collector will implement its own TTL mechanism instead of
+using etcd TTL.
+
+Note that current Pod garbage collection ([capacity-based][]) and Event garbage
+collection ([TTL-based][], using etcd TTL) are not handled by the garbage
+collector. They are kept as-is for backward compatibility. In the future, we may
+move them into garbage collector, or deprecate them and delete them entirely.
+
+[capacity-based]: #non-generic-capacity-based-garbage-collection
+[TTL-based]: #non-generic-ttl-based-garbage-collection
+
+### User Stories
+
+#### Story 1: Clean up old Jobs
+
+The users keep creating Jobs in a small Kubernetes cluster with 4 nodes. The
+Jobs accumulates over time, and 1 year later, the cluster ended up with more
+than 100k old Jobs. This caused etcd hiccups, long high latency etcd requests,
+and eventually made the cluster unavailable.
+
+The problem could have been avoided easily by enabling TTL-based garbage
+collection with Jobs completion time. 
+
+The steps are as easy as:
+
+1. In Job’s metadata, the user sets TTL to 1 hour and that the Job should be
+   cleaned up after it completes (unused).
+1. The user deploys Jobs as usual.
+1. After the Job completes, the result is observed asynchronously within an hour
+   and stored elsewhere. 
+1. The garbage collector cleans up those Jobs 1 hour after Job completion. 
+
+#### Story 2: Secrets and ConfigMaps
+
+Secrets and ConfigMaps are used and referenced by other resources, such as Pods,
+ReplicaSets, and Deployments. An advised pattern is to treat application
+configurations as immutable so that application revisions are intact and
+rollbacks are reliable. When the users deploy their applications, they treat
+Secrets and ConfigMaps as immutable resources, and therefore Secrets and
+ConfigMaps are frequently created and deployed to a Kubernetes cluster, but
+never updated or cleaned up. Secrets and ConfigMaps accumulate over time, which
+could cause master component downtime. 
+
+The problem could have been avoided easily by enabling TTL-based garbage
+collection of Secrets and ConfigMaps by last used time. 
+
+The steps are as easy as:
+
+1. In the ConfigMap and Secret’s metadata, the user sets TTL to 1 hour and that
+   the ConfigMap and Secret should be cleaned up after they are no longer used
+   by other resources (Pods, ReplicaSets, Deployments, etc.)
+1. The user then deploys applications, including ConfigMaps and Secrets of those
+   applications, as usual. Users can create new ConfigMaps and Secrets when they
+   need to change the configuration of their applications. They don’t need to
+   manually check which ConfigMaps and Secrets are not used and can be cleaned
+   up. If the ConfigMaps and Secrets are shared among applications, other
+   applications can still use them before TTL expires. 
```

Comment:
Thanks that's a good example!

Actually I start thinking about combining `.metadata.gcttl` and `.metadata.gcifunused` into one field named `durability` (https://github.com/kubernetes/community/pull/2287#discussion_r197017961), as TTL seems less useful for owner references. An object's `durability` can be `durable`, `ephemeral`, and something else. Each `durability` comes with a sane TTL value.

Reply:
I think the notion of `durability` is hard to be achieved because different applications will have different sane TTL value. +1 for usage reference to handle the case of scaling down to zero.